exercises/concept/expenses/.meta/exemplar.go

```diff
@@ -0,0 +1,127 @@
+package expenses
+
+import (
+	"fmt"
+	"sort"
+	"time"
+)
+
+// Record describes a record of expense event.
+type Record struct {
+	Date     time.Time
+	Amount   float64
+	Category string
+}
+
+// Records describes a collection of expenses events.
+type Records []Record
+
+// Filter creates a new records collection by applying predicate function to
+// collection items and keeping the items when the function returns true.
+func (rr Records) Filter(f func(Record) bool) Records {
+	var out Records
+	for _, r := range rr {
+		if f(r) {
+			out = append(out, r)
+		}
+	}
+	return out
+}
+
+// DatePeriod describes time period.
+type DatePeriod struct {
+	From time.Time
+	To   time.Time
+}
+
+func (p DatePeriod) Includes(d time.Time) bool {
+	return (p.From.Before(d) || p.From.Equal(d)) &&
+		(p.To.After(d) || p.To.Equal(d))
+}
+
+type categoryExpenses struct {
+	Category string
+	Total    float64
+}
+
+type byTotal []categoryExpenses
+
+func (x byTotal) Len() int           { return len(x) }
+func (x byTotal) Less(i, j int) bool { return x[i].Total < x[j].Total }
+func (x byTotal) Swap(i, j int)      { x[i], x[j] = x[j], x[i] }
+
+type byCategoryName []categoryExpenses
+
+func (x byCategoryName) Len() int           { return len(x) }
+func (x byCategoryName) Less(i, j int) bool { return x[i].Category < x[j].Category }
+func (x byCategoryName) Swap(i, j int)      { x[i], x[j] = x[j], x[i] }
+
+func byDatePeriod(p DatePeriod) func(Record) bool {
```

Comment:
I really like the idea of `byDatePeriod` and `byCategory` as good exercises to practice returning functions.

However, I think they should be on their own task. In another comment, I suggested the making of `Filter` also be a task and be the first one, and I think these functions could be the tasks next to that one.

That way, we have these tasks acting as "checkpoints" before the task where we ask the student to write `Total`.